File: sdk/ruby/examples/basic_usage.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# basic_usage.rb — ExperimentationPlatform Ruby SDK usage examples
#
# This file demonstrates the SDK API without making actual HTTP calls.
# For a runnable end-to-end check against a live backend see contract_smoke.rb.

$LOAD_PATH.unshift File.join(__dir__, '..', 'lib')
require 'experimentation_platform'

puts "ExperimentationPlatform Ruby SDK — Basic Usage Examples"
puts "=" * 55
puts

# ---------------------------------------------------------------------------
# 1. Creating a client
# ---------------------------------------------------------------------------
puts "1. Creating a client"
puts "-" * 30

# Option A: keyword arguments
client = ExperimentationPlatform::Client.new(
  base_url:  ENV.fetch("EXPERIMENTLY_API_URL", "http://localhost:8000"),  # origin only
  api_key:   ENV.fetch("EXPERIMENTLY_API_KEY", "sk_live_demo_key"),       # sent as X-API-Key
  cache_ttl: 300,   # cache evaluations/assignments per user + key for 5 minutes
  timeout:   10     # HTTP timeout in seconds
)
puts "   Client created with keyword args"

# Option B: SdkConfig struct
config = ExperimentationPlatform::SdkConfig.new(
  base_url:  "http://localhost:8000",
  api_key:   ENV.fetch("EXPERIMENTLY_API_KEY", "sk_live_demo_key"),
  cache_ttl: 60,
  timeout:   5
)
config.validate!
puts "   SdkConfig created and validated"
puts

# ---------------------------------------------------------------------------
# 2. Evaluating a feature flag (server decides)
# ---------------------------------------------------------------------------
puts "2. Evaluating a feature flag"
puts "-" * 30
puts "   flag = client.evaluate_flag('dark-mode', 'user-123')"
puts "   # GET /api/v1/feature-flags/evaluate/dark-mode?user_id=user-123"
puts "   flag.key       # => 'dark-mode'"
puts "   flag.enabled?  # => true / false (false on any failure)"
puts "   flag.config    # => the flag's config payload, or nil"
puts
puts "   client.feature_enabled?('dark-mode', 'user-123')  # => true / false"
puts

# ---------------------------------------------------------------------------
# 3. Getting an experiment assignment (sticky on the server)
# ---------------------------------------------------------------------------
puts "3. Getting experiment assignment"
puts "-" * 30
puts "   assignment = client.get_assignment('checkout-flow', 'user-123', { country: 'US' })"
puts "   # POST /api/v1/tracking/assign {experiment_key, user_id, context}"
puts "   # nil when the experiment is not ACTIVE or the request fails, otherwise:"
puts "   assignment.experiment_key  # => 'checkout-flow'"
puts "   assignment.variant_name    # => 'control' / 'treatment'"
puts "   assignment.variant_id      # => variant UUID"
puts "   assignment.control?        # => true for the control variant"
puts "   assignment.configuration   # => the variant's configuration Hash, or nil"
puts

# ---------------------------------------------------------------------------
# 4. Tracking events
# ---------------------------------------------------------------------------
puts "4. Tracking events (never raises)"
puts "-" * 30
puts "   # With a key -> one POST /api/v1/tracking/track"
puts "   client.track('purchase', 'user-123', value: 49.0, experiment_key: 'checkout-flow')"
puts "   client.track('search', 'user-123', properties: { q: 'shoes' }, feature_flag_key: 'new-search')"
puts
puts "   # Without a key -> one POST /api/v1/tracking/batch with one entry per cached"
puts "   # assignment and evaluated flag for the user (nothing cached -> nothing sent)"
puts "   client.track('page_view', 'user-123', properties: { page: '/' })"
puts
puts "   # Explicit batch -> BatchResult(success_count, failure_count, errors)"
puts "   client.track_batch(["
puts "     { event_name: 'purchase', user_id: 'user-123', experiment_key: 'checkout-flow', value: 49.0 },"
puts "     { event_name: 'search',   user_id: 'user-123', feature_flag_key: 'new-search' }"
puts "   ])"
puts

# ---------------------------------------------------------------------------
# 5. Consistent hash utility (no HTTP)
# ---------------------------------------------------------------------------
puts "5. Consistent hash utility (golden vector; not used for bucketing)"
puts "-" * 30

bucket = ExperimentationPlatform::FeatureFlagEvaluator.hash_user("user-123", "my-flag")
puts "   hash_user('user-123', 'my-flag') = #{bucket}"
puts "   # Cross-SDK reference value:      0.6927449859213084"
puts "   # Match: #{(bucket - 0.6927449859213084).abs < 1e-10}"
puts

# ---------------------------------------------------------------------------
# 6. Error handling
# ---------------------------------------------------------------------------
puts "6. Error handling"
puts "-" * 30
puts "   # evaluate_flag / get_assignment / track never raise on network or HTTP"
puts "   # errors: they log a warning and return a disabled flag / nil / false."
puts "   # Only HttpClient raises:"
puts "   #   ExperimentationPlatform::AuthenticationError (401)"
puts "   #   ExperimentationPlatform::APIError            (other 4xx/5xx, #status_code)"
puts "   #   ExperimentationPlatform::NetworkError        (timeout, DNS, refused)"
puts

# ---------------------------------------------------------------------------
# 7. Cleanup
# ---------------------------------------------------------------------------
puts "7. Cleanup"
puts "-" * 30
puts "   client.clear_cache  # drop cached results"
puts "   client.close        # same, for shutdown hooks"
puts

client.close

puts "=" * 55
puts "For full API reference see README.md"

File: sdk/php/examples/basic_usage.php
<?php

declare(strict_types=1);

/**
 * Basic usage example for the Experimently PHP SDK.
 *
 * Demonstrates:
 *  1. Creating a client
 *  2. Evaluating a feature flag (decided by the server)
 *  3. Getting a sticky experiment assignment
 *  4. Tracking events (with a key, and key-less fan-out)
 *  5. The cross-SDK hash utility
 *
 * Run from the sdk/php directory:
 *   composer install
 *   EXPERIMENTLY_API_URL=http://localhost:8000 EXPERIMENTLY_API_KEY=... php examples/basic_usage.php
 *
 * For a runnable end-to-end check against a live backend see contract_smoke.php.
 */

require_once __DIR__ . '/../vendor/autoload.php';

use Experimently\ExperimentationClient;
use Experimently\FeatureFlagEvaluator;
use Experimently\SdkConfig;

// ---------------------------------------------------------------------------
// 1. Configure and create the client
// ---------------------------------------------------------------------------

$config = new SdkConfig(
    baseUrl: getenv('EXPERIMENTLY_API_URL') ?: 'http://localhost:8000', // origin only; the SDK appends /api/v1/...
    apiKey: getenv('EXPERIMENTLY_API_KEY') ?: 'your-api-key-here',       // sent as X-API-Key
    cacheTtl: 300,      // reuse a successful evaluation/assignment for 5 minutes
    timeout: 10,        // HTTP timeout in seconds
    maxCacheSize: 1000, // max cached entries
);

$client = new ExperimentationClient($config);

$userId = 'user-' . random_int(1000, 9999);

// ---------------------------------------------------------------------------
// 2. Evaluate a feature flag — GET /api/v1/feature-flags/evaluate/{key}?user_id=...
// ---------------------------------------------------------------------------

echo "=== Feature Flag Evaluation ===\n";

$flag = $client->evaluateFlag('dark-mode', $userId);

echo "Flag:    {$flag->key}\n";
echo "User:    {$userId}\n";
echo 'Enabled: ' . ($flag->enabled ? 'true' : 'false') . "\n";
echo 'Config:  ' . json_encode($flag->config) . "\n";

if ($client->isFeatureEnabled('dark-mode', $userId)) {   // served from the cache
    echo "Dark mode is ENABLED for this user\n";
} else {
    echo "Dark mode is DISABLED (or the flag is not ACTIVE / the API is unreachable)\n";
}

// ---------------------------------------------------------------------------
// 3. Get an experiment assignment — POST /api/v1/tracking/assign (sticky)
// ---------------------------------------------------------------------------

echo "\n=== Experiment Assignment ===\n";

$assignment = $client->getAssignment('checkout-flow-v2', $userId, ['plan' => 'premium', 'country' => 'US']);

if ($assignment !== null) {
    echo "Experiment: {$assignment->experimentKey}\n";
    echo "Variant:    {$assignment->variantName} (id {$assignment->variantId})\n";
    echo 'Is control: ' . ($assignment->isControl ? 'yes' : 'no') . "\n";
    echo 'Config:     ' . json_encode($assignment->configuration) . "\n";
} else {
    echo "No assignment (experiment not ACTIVE, unknown key, or the API is unreachable)\n";
}

// ---------------------------------------------------------------------------
// 4. Track events — never throws
// ---------------------------------------------------------------------------

echo "\n=== Event Tracking ===\n";

// With a key -> one POST /api/v1/tracking/track
$ok = $client->track('purchase', $userId, ['sku' => 'pro-plan'], 'checkout-flow-v2', null, 99.99);
echo 'Tracked purchase (experiment_key): ' . ($ok ? 'success' : 'failed (non-fatal)') . "\n";

$ok = $client->track('search', $userId, ['q' => 'shoes'], null, 'dark-mode');
echo 'Tracked search (feature_flag_key): ' . ($ok ? 'success' : 'failed (non-fatal)') . "\n";

// Without a key -> one POST /api/v1/tracking/batch with one entry per cached
// assignment and evaluated flag for this user (nothing cached -> nothing sent)
$ok = $client->track('page_view', $userId, ['page' => 'checkout']);
echo 'Tracked page_view (fan-out):       ' . ($ok ? 'success' : 'failed (non-fatal)') . "\n";

// Explicit batch -> BatchResult
$result = $client->trackBatch([
    ['event_name' => 'add_to_cart', 'user_id' => $userId, 'experiment_key' => 'checkout-flow-v2', 'value' => 1.0],
    ['event_name' => 'add_to_cart', 'user_id' => $userId, 'feature_flag_key' => 'dark-mode'],
]);
echo "trackBatch: {$result->successCount} ok, {$result->failureCount} failed\n";

// ---------------------------------------------------------------------------
// 5. Cross-SDK hash utility (no network; not used for bucketing)
// ---------------------------------------------------------------------------

echo "\n=== Hash Utility ===\n";

$known = FeatureFlagEvaluator::hashUser('user-123', 'my-flag');
printf("hashUser('user-123', 'my-flag') = %.16f\n", $known);
echo "Expected (all SDKs):              0.6927449859213084\n";
echo 'Match: ' . (abs($known - 0.6927449859213084) < 1e-10 ? 'YES' : 'NO') . "\n";

echo "\nDone.\n";
